File: beijingair.py
from copent import ci
from knncmi import cmi
from causallearn.utils.KCI.KCI import KCI_CInd
from pycit.estimators import ksg_cmi
from fcit import fcit
from CCIT import CCIT
from pcit.IndependenceTest import PCIT  ##  the 'mlxtend' package required

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00381/PRSA_data_2010.1.1-2014.12.31.csv"
prsa2010 = pd.read_csv(url)
# index: 5(PM2.5),6(Dew Point),7(Temperature),8(Pressure),10(Cumulative Wind Speed)
varname = ["","","","","","PM2.5","Dew Point","Temperature","Pressure","","CWS"]
eid = 5 # effect
cid = 8 # cause
data = prsa2010.iloc[2200:2700,[eid,cid]].values

te1 = np.zeros(24)
cmi1 = np.zeros(24)
kci1 = np.zeros(24)
ci1 = np.zeros(24)
fcit1 = np.zeros(24)
ccit1 = np.zeros(24)
pcit1 = np.zeros(24)
#pyc1 = np.zeros(24)
for lag in range(1,25):
	x1 = data[0:(500-lag),0]
	x2 = data[lag:500,0]
	y = data[0:(500-lag),1]
	
	## copent
	te1[lag-1] = ci(x2,y,x1)
	## knncmi
	df = pd.DataFrame(np.vstack((x2,y,x1)).T, columns = ['x2','y','x1'])
	cmi1[lag-1] = cmi(['x2'], ['y'], ['x1'], 3, df)	
	## pycit
	ci1[lag-1] = ksg_cmi(x2, y, x1)
	## fcit
	len1 = len(x1)
	x2a = np.reshape(x2,[len1,1])
	ya = np.reshape(y,[len1,1])
	x1a = np.reshape(x1,[len1,1])
	fcit1[lag-1] = fcit.test(x2a,ya,x1a)
	## causal-learn
	kci_cind = KCI_CInd()
	_,kci1[lag-1] = kci_cind.compute_pvalue(x2a,ya,x1a)
	## CCIT
	ccit1[lag-1] = CCIT.CCIT(x2a,ya,x1a)
	## pcit
	p,tmp,tmp = PCIT(x2a,ya,z = x1a)
	pcit1[lag-1] = p[0]
	
	str = "lag: %d; TE:%.3f; CMI:%.3f; KCI:%.3f; CI:%.3f; fcit:%.3f; ccit:%.3f; pcit:%.3f" %(lag,te1[lag-1],cmi1[lag-1],kci1[lag-1],ci1[lag-1],fcit1[lag-1],ccit1[lag-1],pcit1[lag-1])
	print(str)

data1 = np.vstack((cmi1,kci1,ci1,fcit1,ccit1,pcit1))
df1 = pd.DataFrame(data1.T)
df1.columns = ["cmi","kci","knn","fcit","ccit","pcit"]
df1.to_csv("~/Rworks/bench/py1.csv")

File: beijingair.r
library(corrplot)
library(copent) # transfer entropy (TE)
library(CondIndTests) # kernel-based CI test (KCI)
library(RCIT) # Randomized conditional Correlation Test (RCoT)
library(cdcsis) # conditional distance correlation (CDC)
library(FOCI) # conditional dependence coefficient (CODEC)
library(GeneralisedCovarianceMeasure) # Generalised Covariance Measure (GCM)
library(weightedGCM) # weighted GCM
library(KPC) # Kernel Partial Correlation (KPC)
library(ppcor) # Partial Correlation (pcor)
library(CondCopulas) # Conditional Kendall's Tau (CKT)
library(EDMeasure) # Conditional Mean Dependence (CMD)
source("https://raw.githubusercontent.com/lassepetersen/partial-copula-CI-test/main/parCopCITest.R") # partial copula based CI test

prsa2010data = read.csv("~/Rworks/beijingair/PRSA_data_2010.1.1-2014.12.31.csv")
# id: 6(PM2.5), 7(Dew Point), 8(Temperature), 9(Pressure), 11(Cumulative Wind Speed)
idx = c(6,9)
data = prsa2010data[2200:2700, idx]

ce = 0
# kci = 0
rcot = 0
cdc = 0
codec = 0
gcm = 0
wgcm = 0
kpc = 0
pcor = 0
cmd = 0
pcop = 0

for (lag in 1:24){
  pm25a = data[1:(501-lag),1]
  pm25b = data[(lag+1):501,1]
  v1 = data[1:(501-lag),2]
  
  ce[lag] = ci(pm25b,v1,pm25a)
  # kci[lag] = KCI(pm25b,v1,pm25a)$testStatistic
  rcot[lag] = RCIT(pm25b,v1,pm25a)$Sta
  cdc[lag] = cdcor(pm25b,v1,pm25a)$statistic
  codec[lag] = codec(pm25b,v1,pm25a)
  gcm[lag] = gcm.test(pm25b,v1,pm25a)$test.statistic
  wgcm[lag] = wgcm.est(pm25b,v1,pm25a, regr.meth = "xgboost")
  kpc[lag] = KPCRKHS(pm25b,pm25a,v1)
  pcor[lag] = pcor.test(pm25b,v1,pm25a)$statistic
  cmd[lag] = cmdm_test(pm25b,v1,pm25a)$stat
  pcop[lag] = test_CI(pm25b,v1,pm25a)$statistic
}

# python results
py1 = read.csv("~/Rworks/bench/py1.csv")
kci = py1$kci
cmi1 = py1$knn
cmi2 = py1$cmi
fcit = py1$fcit
ccit = py1$ccit
pcit = py1$pcit

# joint
joint1 = cbind(ce,kci,rcot,cdc,gcm,wgcm,codec,kpc,pcor,cmd,pcop,cmi1,cmi2,fcit,ccit,pcit)
x11()
corrplot(cor(joint1), method = "shade", order = "hclust", col = COL2(n=200))

# plotting
w1 = 5; h1 = 5

# TE via CE
x11(width = w1, height = h1)
plot(ce, xlab = "lag (hours)", ylab = "Transfer Entropy", main = "Pressure")
lines(ce)

# KCI
x11(width = w1, height = h1)
plot(kci, xlab = "lag (hours)", ylab = "KCI", main = "Pressure")
lines(kci)

# RCoT
x11(width = w1, height = h1)
plot(rcot, xlab = "lag (hours)", ylab = "RCoT", main = "Pressure")
lines(rcot)

# CDC
x11(width = w1, height = h1)
plot(cdc, xlab = "lag (hours)", ylab = "CDC", main = "Pressure")
lines(cdc)

# CODEC
x11(width = w1, height = h1)
plot(codec, xlab = "lag (hours)", ylab = "CODEC", main = "Pressure")
lines(codec)

# GCM
x11(width = w1, height = h1)
plot(gcm, xlab = "lag (hours)", ylab = "GCM", main = "Pressure")
lines(gcm)

# wGCM
x11(width = w1, height = h1)
plot(wgcm, xlab = "lag (hours)", ylab = "wGCM", main = "Pressure")
lines(wgcm)

# KPC
x11(width = w1, height = h1)
plot(kpc, xlab = "lag (hours)", ylab = "KPC", main = "Pressure")
lines(kpc)

# Partial Correlation
x11(width = w1, height = h1)
plot(pcor, xlab = "lag (hours)", ylab = "Partial Correlation", main = "Pressure")
lines(pcor)

# Conditional Mean Dependence
x11(width = w1, height = h1)
plot(cmd, xlab = "lag (hours)", ylab = "CMD", main = "Pressure")
lines(cmd)

# partial copula based CI test
x11(width = w1, height = h1)
plot(pcop, xlab = "lag (hours)", ylab = "PartialCopula", main = "Pressure")
lines(pcop)

# cmi1
x11(width = w1, height = h1)
plot(cmi1, xlab = "lag (hours)", ylab = "CMI1", main = "Pressure")
lines(cmi1)

# cmi2
x11(width = w1, height = h1)
plot(cmi2, xlab = "lag (hours)", ylab = "CMI2", main = "Pressure")
lines(cmi2)

# ccit
x11(width = w1, height = h1)
plot(ccit, xlab = "lag (hours)", ylab = "CCIT", main = "Pressure")
lines(ccit)

# fcit
x11(width = w1, height = h1)
plot(fcit, xlab = "lag (hours)", ylab = "FCIT", main = "Pressure")
lines(fcit)

# pcit
x11(width = w1, height = h1)
plot(pcit, xlab = "lag (hours)", ylab = "PCIT", main = "Pressure")
lines(pcit)

